### PR TITLE
Wait for speech adjustment

### PR DESCRIPTION
**Pull request by feerrenrut**
### Link to issue number:
Splitting up PR https://github.com/nvaccess/nvda/pull/14054

### Summary of the issue:
- Some calls to `wait_for_speech_to_finish` do not specify the speech index for speech started (which usually should be `nextSpeechIndex`).
- Generally, it is more concise to use `_NvdaLib.getSpeechAfterKey` rather than `spy.emulateKeyPress`,  `wait_for_speech_to_finish` and then get speech after index. `getSpeechAfterKey` handles this all internally.
- The speech spy synth driver simply called notify for all index commands, and added the speech to an internal buffer, then called doneSpeaking. This doesn't allow the core to process its queue.

### Description of user facing changes
None 

### Description of development approach
- Check all usages of `wait_for_speech_to_finish` and `emulateKeyPress` to see if `getSpeechAfterKey` is more appropriate, or if `nextSpeechIndex` needs to be used.
- The speech spy synth driver now processes speech in a worker thread. Between each utterance, the thread yields to allow core to process any queued callbacks.

### Testing strategy:
System tests on alpha

### Known issues with pull request:
This PR appears to increase the time for tests from 15 min to ~22 min.
- This is based only on the following builds:
   - [21.5 min, build 45160484 ](https://ci.appveyor.com/project/NVAccess/nvda/builds/45160484/messages)
   - [22.1 min, build 45170718](https://ci.appveyor.com/project/NVAccess/nvda/builds/45170718/messages)

### Change log entries:
None

### Code Review Checklist:
- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

File: tests/system/libraries/SystemTestSpy/blockUntilConditionMet.py [hunk 1]
```diff
@@ -12,28 +12,39 @@
...
-bool,  # Was evaluator met?
-Optional[Any]  # None or the value when the evaluator was met
+EvaluatorWasMetT,  # Was evaluator met?
+Optional[GetValueResultT]  # Value when the evaluator was met, if it was met.
 ]:
 	"""Repeatedly tries to get a value up until a time limit expires. Tries are separated by
 	a time interval. The call will block until shouldStopEvaluator returns True when given the value,
 	the default evaluator just returns the value converted to a boolean.
-	@param errorMessage Use 'None' to suppress the exception.
+	@param getValue: Get the value to be tested by shouldStropEvaluator.
```
Comment (seanbudd) [inline tests/system/libraries/SystemTestSpy/blockUntilConditionMet.py:31]: Fix typo

```suggestion
	@param getValue: Get the value to be tested by shouldStopEvaluator.
```

File: tests/system/libraries/SystemTestSpy/blockUntilConditionMet.py [hunk 2]
```diff
@@ -12,28 +12,39 @@
...
+EvaluatorWasMetT = bool
+GetValueResultT = Any
 
 def _blockUntilConditionMet(
-		getValue: Callable[[], Any],
+		getValue: Callable[[], GetValueResultT],
 		giveUpAfterSeconds: float,
 		shouldStopEvaluator=lambda value: bool(value),
-		intervalBetweenSeconds: float = 0.1,
+		intervalBetweenSeconds: float = 0.3,
```
Comment (seanbudd) [inline tests/system/libraries/SystemTestSpy/blockUntilConditionMet.py:22]: Considering this value has been updated across usages, perhaps it should be abstracted to a named constant.

```suggestion
DEFAULT_INTERVAL_BETWEEN_SECONDS = 0.3
...
		intervalBetweenSeconds: float = DEFAULT_INTERVAL_BETWEEN_SECONDS,
```

File: tests/system/libraries/SystemTestSpy/speechSpyGlobalPlugin.py [hunk 3]
```diff
@@ -478,20 +488,24 @@ def _setQueueProcessed():
 				nonlocal queueProcessed
 				queueProcessed = True
 
+			log.debug(f"Waiting for gesture to be processed")
```
Comment (seanbudd) [inline tests/system/libraries/SystemTestSpy/speechSpyGlobalPlugin.py:491]:
```suggestion
			log.debug("Waiting for gesture to be processed")
```

File: tests/system/libraries/SystemTestSpy/speechSpyGlobalPlugin.py [hunk 4]
```diff
@@ -478,20 +488,24 @@ def _setQueueProcessed():
...
 			_blockUntilConditionMet(
 				getValue=lambda: queueProcessed,
 				giveUpAfterSeconds=self._minTimeout(5),
 				errorMessage="Timed out waiting for key to be processed",
 			)
+
 			# We know that by now the core will have woken up and processed the scripts, events and our own function.
 			# Wait for the core to go to sleep,
 			# Which means there is no more things the core is currently processing.
+			log.debug(f"Waiting for core to sleep, to ensure all resulting events have been processed.")
```
Comment (seanbudd) [inline tests/system/libraries/SystemTestSpy/speechSpyGlobalPlugin.py:502]:
```suggestion
			log.debug("Waiting for core to sleep, to ensure all resulting events have been processed.")
```

File: tests/system/libraries/SystemTestSpy/speechSpyGlobalPlugin.py [hunk 5]
```diff
@@ -478,20 +488,24 @@ def _setQueueProcessed():
...
 			_blockUntilConditionMet(
 				getValue=lambda: watchdog.isCoreAsleep(),
 				giveUpAfterSeconds=self._minTimeout(5),
 				errorMessage="Timed out waiting for core to sleep again",
 			)
+			log.debug(f"Core sleeping")
```
Comment (seanbudd) [inline tests/system/libraries/SystemTestSpy/speechSpyGlobalPlugin.py:508]:
```suggestion
			log.debug("Core sleeping")
```

File: tests/system/libraries/SystemTestSpy/speechSpySynthDriver.py [hunk 6]
```diff
@@ -34,21 +56,87 @@ def check(cls):
...
 		synthDriverHandler.synthDoneSpeaking.notify(synth=self)
-		log.debug("post_speech notify start")
+		_yieldThread()
+
+	def _doIndexReached(self, item: IndexCommand):
+		log.debug(f"Speech IndexCommand reached: {item.index}, notifying synthDriverHandler")
+		synthDriverHandler.synthIndexReached.notify(synth=self, index=item.index)
+		_yieldThread()
+
+	def _doNotifySequenceProcessed(self, speechSequence):
```
Comment (seanbudd) [inline tests/system/libraries/SystemTestSpy/speechSpySynthDriver.py:79]:
```suggestion
	def _doNotifySequenceProcessed(self, speechSequence: SpeechSequence):
```

File: tests/system/libraries/SystemTestSpy/speechSpySynthDriver.py [hunk 7]
```diff
@@ -34,21 +56,87 @@ def check(cls):
...
+						timeout=self.POLL_INTERVAL_SECS  # interruptable so that NVDA can exit.
+					)
+				except queue.Empty:
+					if self._speechStarted:
+						self._doDoneSpeaking()
+						self._speechStarted = False
+					continue
+				self._speechStarted = True
+				self._processSpeechSequence(speechSequence)
+		log.debug(f"Stopping")
```
Comment (seanbudd) [inline tests/system/libraries/SystemTestSpy/speechSpySynthDriver.py:102]:
```suggestion
		log.debug("Stopping")
```

File: tests/system/libraries/SystemTestSpy/speechSpySynthDriver.py [hunk 8]
```diff
@@ -34,21 +56,87 @@ def check(cls):
...
+
+		if self._cancel:
+			# stop immediately, don't call done speaking
+			return
+		self._doNotifySequenceProcessed(speechSequence)
 
 	def cancel(self):
-		pass
+		self._cancel = True
+		log.debug(f"Cancelling")
```
Comment (seanbudd) [inline tests/system/libraries/SystemTestSpy/speechSpySynthDriver.py:121]:
```suggestion
		log.debug("Cancelling")
```

File: tests/system/libraries/SystemTestSpy/speechSpySynthDriver.py [hunk 9]
```diff
@@ -34,21 +56,87 @@ def check(cls):
...
+		while not self._queuedSpeech.empty():
+			log.debug(f"discarding.")
```
Comment (seanbudd) [inline tests/system/libraries/SystemTestSpy/speechSpySynthDriver.py:123]:
```suggestion
			log.debug("Discarding queued speech")
```

File: tests/system/libraries/SystemTestSpy/blockUntilConditionMet.py (hunk 2 above)
Comment (feerrenrut) [reply tests/system/libraries/SystemTestSpy/blockUntilConditionMet.py:22]: I've added a constant for `DEFAULT_INTERVAL_BETWEEN_EVAL_SECONDS` and for `_MIN_INTERVAL_BETWEEN_EVAL_SECONDS`, anywhere explicitly passing in `0.3` now uses the default constant.

File: tests/system/libraries/SystemTestSpy/__init__.py [hunk 10]
```diff
@@ -12,6 +12,7 @@
 # Expose shared code. Lint error F401 imported but unused. Exposing like this makes importing easier, and
 # allows code to be re-organized without having to fix many import statements.
 from .blockUntilConditionMet import _blockUntilConditionMet  # noqa: F401
+from .blockUntilConditionMet import DEFAULT_INTERVAL_BETWEEN_EVAL_SECONDS  # noqa: F401
```
Comment (seanbudd) [inline tests/system/libraries/SystemTestSpy/__init__.py:15]: This should probably be

```suggestion
from .blockUntilConditionMet import (  # noqa: F401
	_blockUntilConditionMet,
	DEFAULT_INTERVAL_BETWEEN_EVAL_SECONDS,
)
```

**Review by seanbudd — APPROVED** [review]

**Comment by feerrenrut** [comment]
Last [build test failure](https://ci.appveyor.com/project/NVAccess/nvda/builds/45170718/tests) `Quits from keyboard with welcome dialog open` with error:
```
NVDA failed to exit in the specified timeout
```

The wait time may need to be extended, the test gave up waiting just before NVDA was finished exiting.

<details><summary>Analysis of the failure
</summary>

Highlights from NVDA log (context included below):
<pre><code>
DEBUG - core.triggerNVDAExit (08:01:27.836) - MainThread (7908):
_doShutdown has been queued
...
WARNING - stdout (08:01:28.333) - RF Test Spy Thread (6240):
Robot Framework remote server at 127.0.0.1:8270 stopped.
...
DEBUGWARNING - RPC process 1088 (chrome.exe) (08:01:30.611) - Dummy-4 (7884):
Thread 7756, build\x86_64\remote\ia2Support.cpp, IA2Support_inProcess_terminate, 191:
WaitForMultipleObjects returned 258
...
Terminating speech
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver.cancel (08:01:30.755) - MainThread (7908):
Cancelling
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver._processSpeech (08:01:31.024) - speech spy synth driver (8080):
Stopping
...
</code></pre>
Note the `WaitForMultipleObjects returned 258` which is `WAIT_TIMEOUT 0x00000102L`

Robot log, from after the quit dialog announcement is checked:
<pre><code>
16:01:27.830	INFO	emulateKeyPress ('enter',) {'blockUntilProcessed': False}	
16:01:30.842	FAIL	NVDA failed to exit in the specified timeout	
16:01:30.842	DEBUG	Traceback (most recent call last):
  File "C:\projects\nvda\tests\system\robot\startupShutdownNVDA.py", line 117, in quits_from_keyboard
    errorMessage="NVDA failed to exit in the specified timeout"
  File "C:\projects\nvda\tests\system\libraries\SystemTestSpy\blockUntilConditionMet.py", line 71, in _blockUntilConditionMet
    raise AssertionError(errorMessage)
</code></pre>

NVDA log, from after the quit dialog is announced.
<pre><code>
DEBUG - external:globalPlugins.speechSpyGlobalPlugin.NVDASpyLib.emulateKeyPress (08:01:27.831) - RF Test Spy Thread (6240):
Sending gesture enter
IO - inputCore.InputManager.executeGesture (08:01:27.832) - RF Test Spy Thread (6240):
Input: kb(desktop):enter
DEBUG - core.triggerNVDAExit (08:01:27.836) - MainThread (7908):
_doShutdown has been queued
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (08:01:27.847) - MainThread (7908):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver.cancel (08:01:27.849) - MainThread (7908):
Cancelling
DEBUG - core._terminate (08:01:27.849) - MainThread (7908):
Terminating updateCheck
DEBUG - core._terminate (08:01:27.850) - MainThread (7908):
Terminating watchdog
DEBUG - core._terminate (08:01:27.850) - MainThread (7908):
Terminating globalPluginHandler
DEBUG - external:globalPlugins.speechSpyGlobalPlugin.SystemTestSpyServer.terminate (08:01:27.851) - MainThread (7908):
Terminating the SystemTestSpyServer
DEBUG - braille.BrailleHandler._onBrailleViewerChangedState (08:01:27.852) - MainThread (7908):
Braille Viewer enabled: False
DEBUG - core._closeAllWindows (08:01:27.853) - MainThread (7908):
destroying system tray icon and menu
DEBUG - core._closeAllWindows (08:01:27.856) - MainThread (7908):
closing window <gui.startupDialogs.WelcomeDialog object at 0x067DDB70> during exit process
DEBUG - core._closeAllWindows (08:01:27.857) - MainThread (7908):
closing window <gui.exit.ExitDialog object at 0x067DDE90> during exit process
DEBUG - core._closeAllWindows (08:01:27.869) - MainThread (7908):
destroying main frame during exit process
WARNING - stdout (08:01:28.333) - RF Test Spy Thread (6240):
Robot Framework remote server at 127.0.0.1:8270 stopped.
DEBUGWARNING - RPC process 1088 (chrome.exe) (08:01:30.611) - Dummy-4 (7884):
Thread 7756, build\x86_64\remote\ia2Support.cpp, IA2Support_inProcess_terminate, 191:
WaitForMultipleObjects returned 258

IO - speech.speech.speak (08:01:30.710) - MainThread (7908):
Speaking [LangChangeCommand ('en'), 'NVDA Browser Test Case (1094127420) - Google Chrome', CancellableSpeech (still valid)]
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver.speak (08:01:30.712) - MainThread (7908):
Sequence: [LangChangeCommand ('en'), 'NVDA Browser Test Case (1094127420)   Google Chrome  ', IndexCommand(9)]
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver._processSpeechSequence (08:01:30.713) - speech spy synth driver (8080):
Sequence: [LangChangeCommand ('en'), 'NVDA Browser Test Case (1094127420)   Google Chrome  ', IndexCommand(9)]
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver._doIndexReached (08:01:30.713) - speech spy synth driver (8080):
Speech IndexCommand reached: 9, notifying synthDriverHandler
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver._doNotifySequenceProcessed (08:01:30.714) - speech spy synth driver (8080):
Before notify post_speech
DEBUG - external:globalPlugins.speechSpyGlobalPlugin.NVDASpyLib._onNvdaSpeech (08:01:30.715) - speech spy synth driver (8080):
Appending to speech spy at index 9
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver._doNotifySequenceProcessed (08:01:30.715) - speech spy synth driver (8080):
After post_speech notify
INFO - core.main (08:01:30.722) - MainThread (7908):
Exiting
DEBUG - core.triggerNVDAExit (08:01:30.722) - MainThread (7908):
NVDA has already been triggered to exit safely.
DEBUG - core._terminate (08:01:30.722) - MainThread (7908):
Terminating gui
DEBUG - autoSettingsUtils.autoSettings.AutoSettings._saveSpecificSettings (08:01:30.723) - MainThread (7908):
Saved settings for NVDAHighlighterSettings
DEBUG - autoSettingsUtils.autoSettings.AutoSettings._saveSpecificSettings (08:01:30.723) - MainThread (7908):
Saved settings for ScreenCurtainSettings
DEBUG - fileUtils.FaultTolerantFile (08:01:30.724) - MainThread (7908):
C:\Users\appveyor\AppData\Local\Temp\1\nvdaProfile\nvda.ini6u9231il.tmp
INFO - config.ConfigManager.save (08:01:30.731) - MainThread (7908):
Base configuration saved
DEBUG - core.main (08:01:30.731) - MainThread (7908):
calling lose focus on object with focus
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver.cancel (08:01:30.731) - MainThread (7908):
Cancelling
DEBUG - core._terminate (08:01:30.731) - MainThread (7908):
Terminating treeInterceptorHandler
DEBUG - treeInterceptorHandler.killTreeInterceptor (08:01:30.732) - MainThread (7908):
Killed treeInterceptor: <NVDAObjects.IAccessible.chromium.ChromeVBuf object at 0x065C6B10>
DEBUG - core._terminate (08:01:30.733) - MainThread (7908):
Terminating IAccessible support
DEBUG - core._terminate (08:01:30.734) - MainThread (7908):
Terminating UIA support
DEBUG - core._terminate (08:01:30.744) - MainThread (7908):
Terminating Legacy winConsole support
DEBUG - core._terminate (08:01:30.744) - MainThread (7908):
Terminating Java Access Bridge support
DEBUG - core._terminate (08:01:30.747) - MainThread (7908):
Terminating app module handler
DEBUG - core._terminate (08:01:30.747) - MainThread (7908):
Terminating tones
DEBUG - core._terminate (08:01:30.747) - MainThread (7908):
Terminating NVDAHelper
DEBUG - core._terminate (08:01:30.752) - MainThread (7908):
Terminating touchHandler
DEBUG - core._terminate (08:01:30.752) - MainThread (7908):
Terminating keyboard handler
DEBUG - core._terminate (08:01:30.752) - MainThread (7908):
Terminating mouseHandler
DEBUG - core._terminate (08:01:30.753) - MainThread (7908):
Terminating inputCore
DEBUG - core._terminate (08:01:30.753) - MainThread (7908):
Terminating vision
DEBUG - core._terminate (08:01:30.754) - MainThread (7908):
Terminating brailleInput
DEBUG - core._terminate (08:01:30.754) - MainThread (7908):
Terminating braille
DEBUG - core._terminate (08:01:30.754) - MainThread (7908):
Terminating speech
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver.cancel (08:01:30.755) - MainThread (7908):
Cancelling
DEBUG - external:synthDrivers.speechSpySynthDriver.SpeechSpySynthDriver._processSpeech (08:01:31.024) - speech spy synth driver (8080):
Stopping
DEBUG - core._terminate (08:01:31.024) - MainThread (7908):
Terminating addonHandler
DEBUG - core._terminate (08:01:31.024) - MainThread (7908):
Terminating garbageHandler
WARNING - nvwave.WavePlayer.open (08:01:31.026) - MainThread (7908):
Unable to open WAVE_MAPPER device, there may be no audio devices.
DEBUG - winAPI.sessionTracking.unregister (08:01:31.026) - MainThread (7908):
Unregistered session tracking
DEBUG - core.main (08:01:31.027) - MainThread (7908):
core done
INFO - __main__ (08:01:31.057) - MainThread (7908):
NVDA exit
</code></pre>

</details>

**Pull request merged**